base_layer/core/src/validation/block_body/block_body_partial_validator.rs

```diff
@@ -0,0 +1,321 @@
+
+        check_outputs(constants, &block.body, block.header.height).inspect_err(|e| {
+            error!(
+                target: LOG_TARGET,
+                "BlockBodyPartialValidator: check_outputs height #{}({}): {:?}",
+                block.header.height, block.hash().to_hex(), e
+            )
+        })?;
+
+        verify_no_duplicated_inputs_outputs(&block.body).inspect_err(|e| {
+            error!(
+                target: LOG_TARGET,
+                "BlockBodyPartialValidator: verify_no_duplicated_inputs_outputs {}/{}: {:?}",
+                block.header.height, block.hash().to_hex(), e
+            )
+        })?;
+
+        check_total_burned(&block.body).inspect_err(|e| {
+            error!(
+                target: LOG_TARGET,
+                "BlockBodyPartialValidator: check_total_burned {}/{}: {:?}",
+                block.header.height, block.hash().to_hex(), e
+            )
+        })?;
+
+        verify_timelocks(&block.body, block.header.height).inspect_err(|e| {
+            error!(
+                target: LOG_TARGET,
+                "BlockBodyPartialValidator: verify_timelocks {}/{}: {:?}",
+                block.header.height, block.hash().to_hex(), e
+            )
+        })?;
+
+        Ok(())
+    }
+}
+
+// This function checks that all kernels in the aggregate body exist in the database and that the expected hash
+// corresponds.
+fn validate_kernels_in_db<B: BlockchainBackend>(
+    body: &AggregateBody,
+    db: &B,
+    header_hash: FixedHash,
+) -> Result<(), ValidationError> {
+    for kernel in body.kernels() {
+        match db.fetch_kernel_by_excess_sig(&kernel.excess_sig)? {
+            Some((_kernel, hash)) if hash != header_hash => {
+                let msg = format!(
+                    "'kernels_db' header hash mismatch - expected {}, found {}",
+                    header_hash, hash
+                );
+                return Err(ValidationError::HeaderHashMismatch(msg));
+            },
+            None => {
+                let msg = format!(
+                    "Aggregate body kernel excess {} not found in chain",
+                    kernel.excess.to_hex(),
+                );
+                return Err(ValidationError::MissingKernelError(msg));
+            },
+            _ => {},
+        }
+    }
+    Ok(())
+}
+
+// This function checks that all inputs in the aggregate body exist in the database, either as spent or unspent, and
+// that the expected height and hash corresponds.
+fn validate_input_hashes_in_db<B: BlockchainBackend>(
+    body: &AggregateBody,
+    db: &B,
+    header_hash: FixedHash,
+    header_height: u64,
+) -> Result<(), ValidationError> {
+    let mut inputs: VecDeque<&TransactionInput> = body.inputs().iter().collect();
+    let mut to_remove = Vec::new();
+    for input in &inputs {
+        let input_output_hash = input.output_hash();
+        if db.fetch_output(&input_output_hash)?.is_some() {
+            to_remove.push(input_output_hash);
```

Comment:
this is a problem is not?
this block is in the chain, it as an input, thus the corrosponding output should be marked as spent?